### PR TITLE
Add initial gPodder manifest.

### PR DESCRIPTION
**Pull request by ndarilek**

### CODE REVIEW

File: io.gpodder.gPodder.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
+{
+  "app-id": "io.gpodder.gPodder",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.26",
+  "sdk": "org.gnome.Sdk",
+  "command": "gpodder",
+  "rename-icon": "gpodder",
+  "rename-desktop-file": "gpodder.desktop",
+  "finish-args": [
+    "--socket=x11", "--share=ipc",
```
Comment (nedrichards) [inline io.gpodder.gPodder.json]: does this work in wayland?

File: io.gpodder.gPodder.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    "--socket=system-bus",
```
Comment (nedrichards) [inline io.gpodder.gPodder.json]: which interfaces do you need to talk to? this permission is too broad.

File: io.gpodder.gPodder.json [hunk 3]
```diff
@@ -0,0 +1,77 @@
+{
+  "app-id": "io.gpodder.gPodder",
```
Comment (nedrichards) [inline io.gpodder.gPodder.json]: There doesn't seem to be an actual domain name here. We need the app to comply with the flatpak guidelines: http://docs.flatpak.org/en/latest/introduction.html#identifiers

File: io.gpodder.gPodder.json (hunk 1 above)
Comment (ndarilek) [reply io.gpodder.gPodder.json]: Possibly. I'm blind and Wayland doesn't play nicely with Orca, so I can't confirm.

File: io.gpodder.gPodder.json (hunk 2 above)
Comment (ndarilek) [reply io.gpodder.gPodder.json]: Not sure. The accessibility interfaces at the very least, which I was initially getting errors for on launch. But gPodder needs to communicate with media players and devices, so I'm open to suggestions. New at this, this is my first flatpak, and I haven't done much GNOME app development before so I'm open to suggestions.

**Comment by ndarilek** [comment]
Should be fixed, thanks for the feedback.

File: io.gpodder.gPodder.json (hunk 2 above)
Comment (nedrichards) [reply io.gpodder.gPodder.json]: OK. what you should do is run it with --log-system-bus on the command line and extract the interfaces it wants to talk to and then add them to finish args here. Here's an example of another app owning it's Mpris name on the bus, which is likely something you'll want: https://github.com/flathub/de.haeckerfelix.gradio/blob/3964c87227d35f1e4845a60332d5ceb0fc12af18/de.haeckerfelix.gradio.json#L19

**Comment by nedrichards** [comment]
bot, build io.github.gpodder.gPodder

**Comment by ndarilek** [comment]
So this runs when I do `flatpak-builder --run`, but when I install it and use `flatpak run` it doesn't launch. On the command line I get:

```
Failed to register: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown
```

Also, in `flatpak-builder --run` it uses a former gPodder installation in my home directory, picking up my podcast subscriptions and downloading new episodes. I thought it would be sandboxed. I don't think I give it home directory access anywhere, do I? I'm very confused. I wanted to test this more thoroughly before pushing it, but the inconsistencies make me wonder how effective that is.

Thanks.

**Comment by TingPing** [comment]
Avoid `flatpak-builder --run` it doesn't use the permissions you set.

Run the application with `flatpak run --log-session-bus` to see what permissions you need to add.

File: io.github.gpodder.gPodder.json [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  "app-id": "io.github.gpodder.gPodder",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.26",
+  "sdk": "org.gnome.Sdk",
+  "command": "gpodder",
+  "rename-icon": "gpodder",
+  "rename-desktop-file": "gpodder.desktop",
+  "finish-args": [
+    "--socket=x11", "--share=ipc",
+    "--socket=system-bus",
```
Comment (TingPing) [inline io.github.gpodder.gPodder.json]: This permission is far too broad.

**Comment by ndarilek** [comment]
That's fine, thanks for the pointer. Unfortunately, it doesn't run at
all with `flatpak run`:

```
Failed to register:
GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown:
org.freedesktop.DBus.Error.ServiceUnknown
```

Can't get any help on IRC or the mailing list, so I'm dead in the water.
Interestingly, this did work a time or two, but stopped for reasons of
which I'm uncertain. It also never works from the launcher. Happy to
restrict permissions further, I just need to get it working in order to
do that. :)

Thanks.

**Comment by nedrichards** [comment]
bot, build io.github.gpodder.gPodder

**Comment by TingPing** [comment]
So did you work out the dbus issues or not?

**Comment by ndarilek** [comment]
No. Interestingly enough, it runs in my Atomic Workstation pet Docker
container, but not reliably, and never from the app launcher. I'd really
appreciate some help with this, I'm in over my head. :/ I'm just trying
to square away some of the warnings I get on launch about missing
non-essential modules.

Thanks.

**Comment by TingPing** [comment]
Ok so:

- Rename `app-id` to `org.gpodder.gpodder`
- Add `"--own-name=org.gpodder"` and `"--talk-name=org.gtk.vfs.*"` to finish args

**Comment by ndarilek** [comment]
Thanks! I can make these changes. Would you mind helping me understand
both what they do, and how I might have known to make them? I'd like to
make more flatpaks in the future, and it seems like I might have to up
my debugging game in order to do that. I tried the option to log all
dbus calls but didn't see anything.

**Comment by TingPing** [comment]
Sure this was my process:

```
> flatpak run --log-session-bus io.github.gpodder.gPodder
...
C10: -> org.freedesktop.DBus call org.freedesktop.DBus.RequestName at /org/freedesktop/DBus
Filtering message due to arg0 org.gpodder.gpodder, policy: 0 (required 4)
*HIDDEN* (ping)
....
```

So the application is trying to own that name. If you wonder why: https://github.com/gpodder/gpodder/blob/master/src/gpodder/gtkui/main.py#L3595

So the application sets its own id to `org.gpodder.gpodder`. Changing the flatpak id automatically grants ownership to that name.

So then lets try again:

```
> flatpak run --log-session-bus org.gpodder.gpodder
...
C10: -> org.freedesktop.DBus call org.freedesktop.DBus.NameHasOwner at /org/freedesktop/DBus
Filtering message due to arg0 org.gpodder, policy: 0 (required 1)
B11: <- (no sender) return from C10
*REWRITTEN*
...
...

Well it tries to own `org.gpodder` too. As seen here: https://github.com/gpodder/gpodder/blob/master/share/dbus-1/services/org.gpodder.service.in

Now this is actually more broken because it doesn't match their `app-id` so Flatpak doesn't export its name nor grant permissions to it by default. This should be fixed upstream no matter.

**Comment by ndarilek** [comment]
Thanks, this is great! I was using `--log-system-bus` and seeing
nothing. Like I said, I'm new at this. :) Not sure when something hits
the system bus vs. the session bus.

**Comment by ndarilek** [comment]
Thanks, almost works! I committed this change that doesn't build. I'm
attempting to install the gpod utility to eliminate a startup warning.
Unfortunately, I get:

```
Adding gpod 0.1.2 to easy-install.pth file
error: [Errno 30] Read-only file system:
'/app/lib/python3.5/site-packages/easy-install.pth'
```

I don't understand that, isn't /app/lib specifically *not* read-only as
an installation target?

Thanks again.

**Comment by TingPing** [comment]
`"ensure-writable": ["/lib/python*/site-packages/easy-install.pth"],`

**Comment by ndarilek** [comment]
Thanks. OK, I think this may be ready. The "missing gpod" warnings are
because python-gpod is missing. I just spent a bunch of time trying to
install gpod, but gpod isn't python-gpod which I can't immediately find.

As is, this won't support iPod sync. I hope that's OK for an initial
version and can be added in later. :)

Thanks for all the help.

**Comment by nedrichards** [comment]
bot, build org.gpodder.gpodder

**Comment by ndarilek** [comment]
Cool, so what's the next step with this? I'd like to install it and start using it. I suppose I could install my own copy, but I'm not then sure how I'd uninstall my local build and reinstall the Flathub package without losing all my data.

**Comment by jkozera** [comment]
@ndarilek I'm not the maintainer here, but have you seen https://flathub.org/builds/#/builders/38/builds/26 ? I think the comment above saying

```
"ensure-writable": ["/lib/python*/site-packages/easy-install.pth"],
```

is helpful - sounds like you are missing this part in your file, to allow writing to `easy-install.pth`, which currently fails with:
```
error: [Errno 30] Read-only file system: '/app/lib/python3.5/site-packages/easy-install.pth'
```

(as per the link to the build which I've included)

**Comment by ndarilek** [comment]
Ack, thought I'd committed that change. No, I didn't see the build. Where should I have looked? I'm a blind screen reader user, and am not sure if build status is indicated somewhere in the issue. I'd expect it to appear in the comment stream, but it doesn't seem to.

Thanks, I'll push the change tomorrow.

**Comment by jkozera** [comment]
@ndarilek The builds aren't automatically linked here, I've had to manually check in the Flathub's Buildbot.

**Comment by nedrichards** [comment]
bot, build org.gpodder.gpodder

**Comment by ndarilek** [comment]
Got it, thanks, I'll do that in the future.

Should be fixed now. One minor issue that probably isn't release-critical. I have:

```
  "cleanup": [
    "include"
  ],
```

but there's still ~20K in /app/include. What did I get wrong? I thought the directory was relative to /app unless given absolutely.

**Comment by nedrichards** [comment]
@ndarilek apologies, the buildbot *should* be updating the status here, but it's just not working at the moment. All builds are available on https://flathub.org/builds for future reference. The specific URL is https://flathub.org/builds/#/builders/2/builds/306

**Comment by ndarilek** [comment]
Got it. SO now there's:

```

stat: cannot stat
‘builddir/*/share/app-info/xmls/org.gpodder.gpodder.xml.gz’: No such
file or directory ``` What does this mean? I don't get it locally and it
seems to run.

**Comment by Alexander-Wilms** [comment]
Does it install an appdata.xml file to /app/share/appdata ?

**Comment by ndarilek** [comment]
No, should it? I don't recall reading anything about that.

**Comment by Alexander-Wilms** [comment]
Yes. Many applications install it automatically, but if that's not the case you'll need to create one yourself and copy it to that location.

**Comment by ndarilek** [comment]
bot, build org.gpodder.gpodder

**Comment by nedrichards** [comment]
Everything that your app exports (desktop files, icons, appdata) needs to be named after it's app id. so in this case, org.gpodder.gpodder.appdata.xml you can either put it in /app/share/appdata or /app/share/metainfo (the latter is a more modern path).

The appdata file itself needs to pass the Flathub requirements, which means that yours needs to add OARS rating metadata (fill in the forms at https://hughsie.github.io/oars/ ) and version tags that describe the latest release and it's date (ideally a changelog as well). You can see everything described at the requirements https://github.com/flathub/flathub/wiki/App-Requirements

**Comment by nedrichards** [comment]
bot, build org.gpodder.gpodder

**Comment by nedrichards** [comment]
Build is A-OK now ;-) https://flathub.org/builds/#/builders/2/builds/452 so just a bit more metadata to the appdata file and it'll be good to merge. Thanks very much for your patience through the process.

File: appdata.xml [hunk 5]
```diff
@@ -17,4 +17,36 @@
...
+    <content_attribute id="social-chat">none</content_attribute>
+    <content_attribute id="social-info">none</content_attribute>
+    <content_attribute id="social-audio">none</content_attribute>
+    <content_attribute id="social-location">none</content_attribute>
+    <content_attribute id="social-contacts">none</content_attribute>
+    <content_attribute id="money-purchasing">none</content_attribute>
+    <content_attribute id="money-gambling">none</content_attribute>
+  </content_rating>
+  <releases>
+    <release version="3.10.1"/>
```
Comment (nedrichards) [inline appdata.xml]: I think this also needs a date property as per https://github.com/flathub/org.geogebra.GeoGebra/blob/master/appdata.xml#L19 but I'll build it anyway - if it's fine we can merge and you can add it yourself to the repo later.

**Comment by nedrichards** [comment]
bot, build org.gpodder.gpodder

File: appdata.xml (hunk 5 above)
Comment (nedrichards) [reply appdata.xml]: > org.gpodder.gpodder:               AppData problem: attribute-missing : <release> has no timestamp

As expected the build fails https://flathub.org/builds/#/builders/5/builds/58

**Comment by nedrichards** [comment]
bot, build org.gpodder.gpodder

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/org.gpodder.gpodder

**Pull request closed without merging**